File: python/mirage/core/object_store/du.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.cache.index import NULL_INDEX, IndexCacheStore
from mirage.core.object_store.driver import (A, C, DuEntriesFn, DuSizeFn,
                                             ObjectStoreDriver)
from mirage.core.object_store.readdir import read_tree
from mirage.types import PathSpec
from mirage.utils import key_prefix as kp


def make_du_entries(driver: ObjectStoreDriver[A, C]) -> DuEntriesFn[A]:
    """Build the per-object size walk over one driver.

    Keys are stripped back to mount-relative paths, so a store mounted
    at a ``key_prefix`` reports the paths the user typed rather than the
    raw keys.

    Args:
        driver (ObjectStoreDriver): the store's native surface.
    """

    async def entries(accessor: A,
                      path_spec: PathSpec,
                      index: IndexCacheStore = NULL_INDEX
                      ) -> tuple[list[tuple[str, int]], int]:
        kpfx = driver.key_prefix_of(accessor)
        found: list[tuple[str, int]] = []
        total = 0
        rows, _ = await read_tree(driver, accessor, path_spec, index)
        for entry in rows:
            if not entry.key or entry.key.endswith("/"):
                continue
            rel = kp.strip(kpfx, entry.key)
            found.append(("/" + rel.lstrip("/"), entry.size or 0))
            total += entry.size or 0
        found.sort()
        return found, total

    return entries


def make_du_size(driver: ObjectStoreDriver[A, C]) -> DuSizeFn[A]:
    """Build the recursive byte total over one driver.

    Args:
        driver (ObjectStoreDriver): the store's native surface.
    """

    async def size(accessor: A,
                   path_spec: PathSpec,
                   index: IndexCacheStore = NULL_INDEX) -> int:
        rows, _ = await read_tree(driver, accessor, path_spec, index)
        return sum(entry.size or 0 for entry in rows
                   if entry.key and not entry.key.endswith("/"))

    return size

File: typescript/packages/core/src/core/object_store/du.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { Accessor } from '../../accessor/base.ts'
import * as kp from '../../utils/key_prefix.ts'
import { lstripSlash } from '../../utils/slash.ts'
import { compareCodePoints } from '../../utils/sort.ts'
import { readTree } from './readdir.ts'
import type { DuEntriesFn, DuSizeFn, ObjectStoreDriver } from './driver.ts'

/**
 * Build the per-object size walk over one driver.
 *
 * Keys are stripped back to mount-relative paths, so a store mounted at
 * a `keyPrefix` reports the paths the user typed rather than the raw
 * keys.
 */
export function makeDuEntries<A extends Accessor, C>(
  driver: ObjectStoreDriver<A, C>,
): DuEntriesFn<A> {
  return async function entries(accessor, path, index) {
    const kpfx = driver.keyPrefixOf(accessor)
    const found: [string, number][] = []
    let total = 0
    const [rows] = await readTree(driver, accessor, path, index)
    for (const entry of rows) {
      if (entry.key === '' || entry.key.endsWith('/')) continue
      const rel = kp.strip(kpfx, entry.key)
      const size = entry.size ?? 0
      found.push(['/' + lstripSlash(rel), size])
      total += size
    }
    found.sort((a, b) => compareCodePoints(a[0], b[0]) || a[1] - b[1])
    return [found, total]
  }
}

/** Build the recursive byte total over one driver. */
export function makeDuSize<A extends Accessor, C>(driver: ObjectStoreDriver<A, C>): DuSizeFn<A> {
  return async function size(accessor, path, index) {
    const [rows] = await readTree(driver, accessor, path, index)
    return rows.reduce(
      (total, row) => total + (row.key !== '' && !row.key.endsWith('/') ? (row.size ?? 0) : 0),
      0,
    )
  }
}
